pkg/api/validation/validation.go

```diff
@@ -2518,59 +2531,90 @@ func validateServicePort(sp *api.ServicePort, requireName, isHeadlessService boo
+			"can only be set on loadBalancer service with ExternalTraffic=OnlyLocal"))
+	}
+	return allErrs
+}
 
-	_, healthCheckBetaOk := service.Annotations[apiservice.BetaAnnotationHealthCheckNodePort]
-	_, onlyLocalBetaOk := service.Annotations[apiservice.BetaAnnotationExternalTraffic]
+type serviceExternalTrafficAnnotationsUpdateStatus struct {
+	createBeta    bool
+	updateBeta    bool
+	deleteBeta    bool
+	useBetaWithGA bool
+}
 
-	var oldHealthCheckAlpha, oldOnlyLocalAlpha string
-	var oldHealthCheckAlphaOk, oldOnlyLocalAlphaOk bool
+func (s *serviceExternalTrafficAnnotationsUpdateStatus) checkStatus(newService, oldService *api.Service, annotationKey string,
+	GASelfIsSet, GAPeerIsSet bool) {
+	// Collect status from the new service.
+	betaValue, betaIsSet := newService.Annotations[annotationKey]
+
+	// Collect status from the old service.
+	var oldBetaValue string
+	var oldBetaIsSet bool
 	if oldService != nil {
-		oldHealthCheckAlpha, oldHealthCheckAlphaOk = oldService.Annotations[apiservice.AlphaAnnotationHealthCheckNodePort]
-		oldOnlyLocalAlpha, oldOnlyLocalAlphaOk = oldService.Annotations[apiservice.AlphaAnnotationExternalTraffic]
-	}
-	hcValueChanged := oldHealthCheckAlphaOk && healthCheckAlphaOk && oldHealthCheckAlpha != hcAlpha
-	hcValueNew := !oldHealthCheckAlphaOk && healthCheckAlphaOk
-	hcValueGone := !healthCheckAlphaOk && !healthCheckBetaOk && oldHealthCheckAlphaOk
-	onlyLocalHCMismatch := onlyLocalBetaOk && healthCheckAlphaOk
-
-	// On upgrading to a 1.5 cluster, the user is locked in at the current
-	// alpha setting, till they modify the Service such that the pair of
-	// annotations are both beta. Basically this means we need to:
-	// Disallow updates to the alpha annotation.
-	// Disallow creating a Service with the alpha annotation.
-	// Disallow removing both alpha annotations. Removing the health-check
+		oldBetaValue, oldBetaIsSet = oldService.Annotations[annotationKey]
+	}
+
+	// Flip the flags for all invalid changes.
+	s.createBeta = betaIsSet && !oldBetaIsSet
+	s.updateBeta = betaIsSet && oldBetaIsSet && betaValue != oldBetaValue
+	s.deleteBeta = !GASelfIsSet && !betaIsSet && oldBetaIsSet
+	s.useBetaWithGA = betaIsSet && GAPeerIsSet
+}
+
+func (s *serviceExternalTrafficAnnotationsUpdateStatus) invalidChange() bool {
+	return s.createBeta || s.updateBeta || s.deleteBeta || s.useBetaWithGA
+}
+
+func validateServiceExternalTrafficAnnotationsUpdate(service, oldService *api.Service) (allErrs field.ErrorList) {
+	// 2 annotations went from beta to GA in 1.6: healthcheck-nodeport and
+	// external-traffic. The user cannot mix these. All updates to the beta
+	// annotation are disallowed. The user must change both beta annotations
+	// to first class fields before making any modifications, even though the
+	// system continues to respect the beta version.
+
+	// On upgrading to a 1.6 cluster, the user is locked in at the current
```

Comment:
I wanted to say up-conversion, but saw [your old comment about it](https://github.com/kubernetes/kubernetes/pull/33957#discussion_r82726939). Supporting beta annotations for one more release seems to be the right thing to do. I will re-think this part.